### PR TITLE
Add pipeline versions table, refactor pipelines table and pipelines selector

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
JIRA:
https://issues.redhat.com/browse/RHOAIENG-263
https://issues.redhat.com/browse/RHOAIENG-264

## Description

The PR did the following things:
1. Refactor the global pipelines table, and change it to bulk selection, change the columns to get the versions number and updated date (the latest version date)

Screenshot 2023-12-14 at 4 39 55 PM

2. Add the pipeline versions table to the expandable section in the row of the pipeline, when it's more than 10 versions, you can click on the View More button to load more versions, you can also sort the versions by name and create time.

**NOTE: The view run Button link is not implemented yet, needs to be addressed when refactoring the pipeline runs page.**

Screenshot 2023-12-14 at 4 42 16 PM

3. Refactor the pipeline and pipeline version selector to make it also use the server-side filtering and sorting to keep the behavior aligned with what we have now.

## How Has This Been Tested?

1. Go to the global pipelines page
2. Try to expand the pipelines to see the pipeline versions
3. Click on the version should be able to navigate you to the version details page
4. Try to upload new versions to the pipelines
5. Try to create more than 10 pipeline versions under a pipeline to test the view more function
6. Try to sort the pipelines table and the pipeline versions table
7. Try to delete all the versions in one pipeline to check this state
Screenshot 2023-12-14 at 4 47 30 PM
8. Play around with the pipeline selectors, it should have better filtering and sorting now

## Test Impact

Update the test cases for the pipeline selector. Didn't create new tests, because it makes more sense to use e2e tests for this.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Comment by DaoDaoNoCode** [comment]
@yannnz Could you check the UI on this page?

**Review by jpuzz0 — COMMENTED** [review]
In general, it seems like there's a lot being introduced here that could use unit testing at least (and at least for new hooks/utils). I know you mentioned e2e tests to be added later and some other refactoring in the description, and we should probably track that work now so it is not lost in mix.

File: frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      {},
+      {
+        pageToken: pageTokenRef.current,
+        pageSize: 10,
+        sortField,
+        sortDirection,
+        filter,
+      },
+    );
+    setMoreData(result.pipelines || []);
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts:57]: Yes, otherwise this won't re-render. This is a work-around to trigger re-render.

File: frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      pageToken: pageTokenRef.current,
+      pageSize: 10,
+      sortField,
+      sortDirection,
+      filter,
+    });
+    setMoreData(result.versions || []);
+    dataRef.current = [...dataRef.current, ...(result.versions || [])];
+    pageTokenRef.current = result.next_page_token;
+    return undefined;
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts:102]: Because it will fail the type check if we don't add it
> Not all code paths return a value.
We use if statement above to check the pipeline id and return a Promise reject above.

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          }
+          style={{ maxWidth, minWidth }}
+          ref={toggleRef}
+          onClick={() => setOpen(!isOpen)}
+          isExpanded={isOpen}
+          isDisabled={!pipelineId || totalSize === 0}
+          isFullWidth
+          data-testid="pipeline-version-toggle-button"
+        >
+          {!pipelineId
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:147]: Yes, this is an agreement by me and @christianvogt. When there is no pipeline id, all we want to remind the user is to choose a pipeline first (on Pipeline runs page)

File: frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts:57]: Also, instances like this, if there is no comment associated with it explaining its intention, its difficult to assess its purpose, unless the naming of the variable can be more explanatory.

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:147]: Right, but this component is independent of the pipeline runs page and is used in different places where that might not make sense (anywehere that doesn't have a pipeline input right next to this selector).

[the remark this answers is not in the document]
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:147]: I'm not sure if this makes sense to check for a pipeline ID and show a message at all for that condition, but if we absolutely want to check for a pipelineId, I assume if 1 is not found then "no versions available" would make more sense, since there would therefore be no versions returned from the API response?

File: frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts [hunk 4]
```diff
@@ -96,9 +96,10 @@ export const usePipelineVersionLoadMore = ({
       sortDirection,
       filter,
     });
-    setMoreData(result.versions || []);
+    showMoreData((flag) => !flag);
     dataRef.current = [...dataRef.current, ...(result.versions || [])];
     pageTokenRef.current = result.next_page_token;
+    // Return undefined to avoid type check error because there is a if return above
```
Comment (jpuzz0) [inline frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts:102]: I don't think this comment is necessary, but rather one explaining the purpose of `showMoreData` above.

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:147]: I can change that, though it's on the mockup saying `Select a pipeline version`

Screenshot 2023-12-15 at 12 44 22 PM

**Comment by jpuzz0** [comment]
Tested this locally, and functionally it seems fine. I think some of these newly created hooks/functions, especially some of the complicated ones should have some unit testing, ideally before merging, but otherwise tracked in a separate ticket as a fast-follow.

File: frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  sortField,
+  loaded,
+  filter,
+}: PipelineLoadMoreProps) => {
+  const { api } = usePipelinesAPI();
+  const [, setMoreData] = React.useState<PipelineKF[]>([]);
+  const dataRef = React.useRef<PipelineKF[]>([]);
+  const pageTokenRef = React.useRef<string>();
+
+  React.useMemo(() => {
```
[the remark this answers is not in the document]
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/content/tables/usePipelineLoadMore.ts:37]: Spoke with Juntao on slack and he explained the reasoning behind the logic here.

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpuzz0) [reply frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:147]: I think the placeholder logic should be to show "Select a pipeline version" by default, and on this page, if no pipeline is selected (and therefore no pipelineId passed to the selector for the version), the input would be disabled.

File: frontend/src/concepts/pipelines/content/tables/pipelineVersion/PipelineVersionTable.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        itemCount={totalSize}
+        data={versions}
+        columns={pipelineVersionColumns}
+        rowRenderer={(version) => (
+          <PipelineVersionTableRow
+            key={version.id}
+            isChecked={isSelected(version.id)}
+            onToggleCheck={() => toggleSelection(version.id)}
+            version={version}
+            pipelineVersionDetailsPath={pipelineVersionDetailsPath}
```
Comment (christianvogt) [inline frontend/src/concepts/pipelines/content/tables/pipelineVersion/PipelineVersionTable.tsx:62]: Is it necessary to pass along this prop? Is there another use case where the link path will differ? If not just generate the link path inside `PipelineVersionTableRow`

File: frontend/src/concepts/pipelines/content/tables/pipeline/usePipelineTableRowData.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+const usePipelineTableRowData = (pipeline: PipelineKF) => {
+  const [{ items, totalSize }, isLoaded, , refresh] = usePipelineVersionsForPipeline(pipeline.id, {
+    pageSize: 1,
+    sortField: 'created_at',
+    sortDirection: 'desc',
+  });
+
+  const updatedDate = new Date(
+    items.length === 0 || !items[0] ? pipeline.created_at : items[0].created_at,
```
Comment (christianvogt) [inline frontend/src/concepts/pipelines/content/tables/pipeline/usePipelineTableRowData.ts:12]:
```suggestion
    items?.[0]?.created_at || pipeline.created_at,
```

File: frontend/src/concepts/pipelines/content/pipelineSelector/utils.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  const search = getDataValue(filterData[FilterOptions.NAME]);
+
+  const totalSize = React.useMemo(() => {
+    if (loaded && !search) {
+      totalSizeRef.current = fetchedSize;
+      return fetchedSize;
+    }
+    return totalSizeRef.current;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [loaded]);
```
Comment (christianvogt) [inline frontend/src/concepts/pipelines/content/pipelineSelector/utils.ts:68]: Please always add comments when disabling an eslint rule. In particular this memo solves a specific use case where we need to catch the total size.

File: frontend/src/concepts/pipelines/content/createRun/contentSections/PipelineSection.tsx [hunk 9]
```diff
@@ -5,62 +5,39 @@ import {
...
+const PipelineSection: React.FC<PipelineSectionProps> = ({ value, onChange }) => (
+  <FormSection
+    id={CreateRunPageSections.PIPELINE}
+    title={runPageSectionTitles[CreateRunPageSections.PIPELINE]}
+  >
+    <Stack hasGutter>
+      <StackItem>
+        <PipelineSelector
+          selection={value?.name}
+          maxWidth="500px"
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/createRun/contentSections/PipelineSection.tsx:26]: Yes, I don't know why it's not respecting the form max width, but it is 🤷 

https://github.com/opendatahub-io/odh-dashboard/assets/37624318/a3a276e0-4965-4454-8c09-7ad5e58317d8

File: frontend/src/concepts/pipelines/content/tables/pipelineVersion/PipelineVersionTable.tsx (hunk 6 above)
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/tables/pipelineVersion/PipelineVersionTable.tsx:62]: I think the projects details page also needs this, let me double check it, maybe I forgot to take care of that page.

File: frontend/src/concepts/pipelines/content/import/PipelineVersionImportModal.tsx [hunk 10]
```diff
@@ -151,4 +137,8 @@ const PipelineVersionImportModal: React.FC<PipelineVersionImportModalProps> = ({
   );
 };
 
-export default PipelineVersionImportModal;
+const PipelineVersionImportModalWrapper = (props: PipelineVersionImportModalProps) => (
+  <PipelineVersionImportModal key={props.existingPipeline?.id} {...props} />
+);
+
+export default PipelineVersionImportModalWrapper;
```
Comment (christianvogt) [inline frontend/src/concepts/pipelines/content/import/PipelineVersionImportModal.tsx:144]: Why is this needed?

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        <MenuSearch>
+          <MenuSearchInput>
+            <SearchInput {...searchProps} aria-label="Filter pipeline versions" />
+          </MenuSearchInput>
+          <HelperText>
+            <HelperTextItem variant="indeterminate">{`Type a name to search your ${totalSize} versions.`}</HelperTextItem>
+          </HelperText>
+        </MenuSearch>
+        <MenuList>
+          <div role="menuitem">
```
Comment (christianvogt) [inline frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:78]: Can this not be a PF `MenuItem` ?

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            <TableBase
+              {...tableProps}
+              itemCount={fetchedSize}
+              loading={!loaded}
+              data-id="pipeline-version-selector-table-list"
+              emptyTableView={
+                <EmptyTableView
+                  hasIcon={false}
```
Comment (christianvogt) [inline frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:86]:
```suggestion
```

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+                  variant={EmptyStateVariant.xs}
+                />
+              }
+              borders={false}
+              variant={TableVariant.compact}
+              columns={pipelineVersionSelectorColumns}
+              data={versions}
+              rowRenderer={(row, index) => (
+                <PipelineSelectorTableRow
+                  key={index}
```
Comment (christianvogt) [inline frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:97]: Use `row.id` (or something else) instead of `index`.
```suggestion
                  key={row.id}
```

File: frontend/src/concepts/pipelines/content/pipelineSelector/utils.ts [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            direction: sortDirection,
+            defaultDirection: 'asc',
+          },
+          onSort: (_event, index, direction) => {
+            setSortField(String(columns[index].field));
+            setSortDirection(direction);
+          },
+          columnIndex,
+        }
+      : undefined;
```
Comment (christianvogt) [inline frontend/src/concepts/pipelines/content/pipelineSelector/utils.ts:40]: Similar code appears 7 times in the code base. It would be good to remove all instances into a single configurable function.

File: frontend/src/concepts/pipelines/content/import/PipelineVersionImportModal.tsx (hunk 10 above)
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/import/PipelineVersionImportModal.tsx:144]: On the create pipeline runs page, if you change the pipeline selection and click the `Upload new version` button, it will not update the pipeline selection in the pipeline version import modal based on the pipeline selection. This way could guarantee the import versions modal re-renders every time when the pipeline changes.

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx (hunk 11 above)
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:78]: I remember there are some styles that I don't want as a MenuItem. But I need the role to get rid of the warning from the console.

File: frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx (hunk 12 above)
Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:86]: Why do we want to remove this? I figured out that the empty state in the selector could be very tiny so we might not want an icon for it. (Followed the extra small empty state example here https://www.patternfly.org/components/empty-state#extra-small)

Comment (christianvogt) [reply frontend/src/concepts/pipelines/content/pipelineSelector/PipelineVersionSelector.tsx:86]: @DaoDaoNoCode oh i thought that `hasIcon = false` was the default is all. If it's not the default then certainly keep it.

File: frontend/src/concepts/pipelines/content/import/PipelineVersionImportModal.tsx (hunk 10 above)
Comment (christianvogt) [reply frontend/src/concepts/pipelines/content/import/PipelineVersionImportModal.tsx:144]: I think this modal is done incorrectly because even if we do not yet open the modal, it has potential for fetching information. We want to avoid doing work for the modal while it it closed. It's likely copying a bad pattern used across the entire app ...

Can we change this modal so that it only does real work when it is opened. I suspect that would solve your issue and remove the need for this `key` trick.

Comment (DaoDaoNoCode) [reply frontend/src/concepts/pipelines/content/import/PipelineVersionImportModal.tsx:144]: Yes, this modal state setting is incorrect across the app lol... Let me see how can I solve this.

**Comment by DaoDaoNoCode** [comment]
[RHOAIENG-1124](https://issues.redhat.com/browse/RHOAIENG-1124) for adding tests.
